### PR TITLE
[SPARK-47487][SQL] Simplify code in AnsiTypeCoercion

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

Simplify the code in `AnsiTypeCoercion.implicitCast`, to merge common code paths.

### Why are the changes needed?

improve code readability

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

existing tests

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala [hunk 1]
```diff
@@ -180,37 +180,18 @@ object AnsiTypeCoercion extends TypeCoercionBase {
       // cast the input to decimal.
       case (n: NumericType, DecimalType) => Some(DecimalType.forType(n))
 
-      // Cast null type (usually from null literals) into target types
-      // By default, the result type is `target.defaultConcreteType`. When the target type is
-      // `TypeCollection`, there is another branch to find the "closet convertible data type" below.
-      case (NullType, target) if !target.isInstanceOf[TypeCollection] =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala:186]: This is covered by `Cast.canANSIStoreAssign`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala [hunk 2]
```diff
@@ -180,37 +180,18 @@ object AnsiTypeCoercion extends TypeCoercionBase {
...
-      // Decimal type.
-      case (StringType, DecimalType) =>
-        Some(DecimalType.SYSTEM_DEFAULT)
-
-      // If the target type is any timestamp type, convert the String type as the default
-      // Timestamp type.
-      case (StringType, AnyTimestampType) =>
-        Some(AnyTimestampType.defaultConcreteType)
-
-      case (DateType, AnyTimestampType) =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala:208]: ditto

**Comment by cloud-fan** [comment]
cc @gengliangwang @MaxGekk

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala [hunk 3]
```diff
@@ -180,37 +180,18 @@ object AnsiTypeCoercion extends TypeCoercionBase {
...
-        Some(AnyTimestampType.defaultConcreteType)
-
-      case (_, target: DataType) =>
-        if (Cast.canANSIStoreAssign(inType, target)) {
-          Some(target)
+      // Ideally the implicit cast rule should be the same as `Cast.canANSIStoreAssign` so that it's
+      // consistent with table insertion. To avoid breaking too many existing Spark SQL queries,
+      // we make the system to allow implicitly converting String type as other primitive types.
+      case (StringType, a @ (_: AtomicType | NumericType | DecimalType | AnyTimestampType)) =>
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala:186]: @cloud-fan Isn't  `NumericType` a sub-class of `AtomicType`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala (hunk 1 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala:186]:
```
      case (_, target: DataType) =>
        if (Cast.canANSIStoreAssign(inType, target)) {
          Some(target)
        } else {
          None
        }
```

There seems to be some differences between returning `target` and `target.defaultConcreteType`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala [hunk 4]
```diff
@@ -180,56 +180,27 @@ object AnsiTypeCoercion extends TypeCoercionBase {
...
+        Some(a.defaultConcreteType)
+
+      // When the target type is `TypeCollection`, there is another branch to find the
+      // "closet convertible data type" below.
+      case (_, target) if !target.isInstanceOf[TypeCollection] =>
+        val concreteType = target.defaultConcreteType
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala:197]: Here you are making an assumption that each `DataType`'s defaultConcreteType is itself.
```
case (_, target: DataType) =>
        if (Cast.canANSIStoreAssign(inType, target)) {
          Some(target)
        } else {
          None
        }
```
Shall we be more conservative and make it
```
val concreteType = target match {
case d: DataType => d
case _ => target.defaultConcreteType
}
```
?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala:197]: It's the default implementation in `DataType`:
```
override private[sql] def defaultConcreteType: DataType = this
```

maybe I just make it as `final`?

Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala:197]: Making it as `final` also sounds good.

**Review by gengliangwang — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]

**Review by yaooqinn — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Pull request closed without merging**